src/lektorium/client/src/auth/authService.js

```diff
@@ -0,0 +1,140 @@
+
+  const localStorageKey = "loggedIn";
+  const loginEvent = "loginEvent";
+
+  class AuthService extends EventEmitter {
+    idToken = null;
+    accessToken = null;
+    profile = null;
+    tokenExpiry = null;
+
+    login(customState) {
+      webAuth.authorize({
+        appState: customState
+      });
+    }
+
+    logOut() {
+      localStorage.removeItem(localStorageKey);
+
+      this.idToken = null;
+      this.accessToken = null;
+      this.tokenExpiry = null;
+      this.profile = null;
+
+      webAuth.logout({
+        returnTo: `${window.location.origin}`
+      });
+
+      this.emit(loginEvent, { loggedIn: false });
+    }
+
+    handleAuthentication() {
+      return new Promise((resolve, reject) => {
+        webAuth.parseHash((err, authResult) => {
+          if (err) {
+            this.emit(loginEvent, {
+              loggedIn: false,
+              error: err,
+              errorMsg: err.statusText
+            });
+            reject(err);
+          } else {
+            this.localLogin(authResult);
+            resolve(authResult.idToken);
+          }
+        });
+      });
+    }
+
+    isAuthenticated() {
+      return (
+        Date.now() < this.tokenExpiry &&
+        localStorage.getItem(localStorageKey) === "true"
+      );
+    }
+
+    isTokenValid() {
+      return this.idToken && this.tokenExpiry && Date.now() < this.tokenExpiry;
+    }
+
+    getTokens() {
+      return new Promise((resolve, reject) => {
+        if (this.isTokenValid()) {
+          resolve([this.idToken, this.accessToken]);
+        } else if (this.isAuthenticated()) {
+          this.renewTokens().then(authResult => {
+            resolve([authResult.idToken, authResult.accessToken]);
+          }, reject);
+        } else {
+          resolve();
+        }
+      });
+    }
+
+    localLogin(authResult) {
+      var profile = JSON.parse(JSON.stringify(authResult.idTokenPayload));
+      profile.access_token = parseJwt(authResult.accessToken);
+      this.idToken = authResult.idToken;
+      this.accessToken = authResult.accessToken;
+      this.profile = profile;
+
+      // Convert the expiry time from seconds to milliseconds,
+      // required by the Date constructor
+      this.tokenExpiry = new Date(this.profile.exp * 1000);
+
+      localStorage.setItem(localStorageKey, "true");
+
+      this.emit(loginEvent, {
+        loggedIn: true,
+        profile: profile,
+        state: authResult.appState || {}
+      });
+    }
+
+    renewTokens() {
+      return new Promise((resolve, reject) => {
+        if (localStorage.getItem(localStorageKey) !== "true") {
+          return reject("Not logged in");
+        }
+
+        webAuth.checkSession({}, (err, authResult) => {
+          if (err) {
+            reject(err);
+          } else {
+            this.localLogin(authResult);
+            resolve(authResult);
+          }
+        });
+      });
+    }
+  }
+
+  const service = new AuthService();
+  service.setMaxListeners(5);
```

Comment:
Why 5?

Reply:
It is copy/paste from auth0 js example. It is working and I do not want to touch it.